truffle/src/main/java/org/jruby/truffle/format/parser/PrintfTreeBuilder.java

```diff
@@ -0,0 +1,240 @@
+/*
+ * Copyright (c) 2015 Oracle and/or its affiliates. All rights reserved. This
+ * code is released under a tri EPL/GPL/LGPL license. You can use it,
+ * redistribute it and/or modify it under the terms of the:
+ *
+ * Eclipse Public License version 1.0
+ * GNU General Public License version 2
+ * GNU Lesser General Public License version 2.1
+ */
+package org.jruby.truffle.format.parser;
+
+import com.oracle.truffle.api.object.DynamicObject;
+import org.antlr.v4.runtime.Token;
+import org.antlr.v4.runtime.misc.Interval;
+import org.jruby.truffle.format.nodes.PackNode;
+import org.jruby.truffle.format.nodes.SourceNode;
+import org.jruby.truffle.format.nodes.control.SequenceNode;
+import org.jruby.truffle.format.nodes.format.FormatFloatNodeGen;
+import org.jruby.truffle.format.nodes.format.FormatIntegerNodeGen;
+import org.jruby.truffle.format.nodes.read.*;
+import org.jruby.truffle.format.nodes.type.ToDoubleWithCoercionNodeGen;
+import org.jruby.truffle.format.nodes.type.ToIntegerNodeGen;
+import org.jruby.truffle.format.nodes.type.ToStringNodeGen;
+import org.jruby.truffle.format.nodes.write.WriteByteNode;
+import org.jruby.truffle.format.nodes.write.WriteBytesNodeGen;
+import org.jruby.truffle.format.nodes.write.WritePaddedBytesNodeGen;
+import org.jruby.truffle.runtime.RubyContext;
+import org.jruby.truffle.runtime.layouts.Layouts;
+import org.jruby.util.ByteList;
+import org.jruby.util.StringSupport;
+
+import java.util.ArrayList;
+import java.util.List;
+
+public class PrintfTreeBuilder extends org.jruby.truffle.format.parser.PrintfParserBaseListener {
+
+    // TODO CS 19-Dec-15 this is never used, but the functionality seems useful
+    public static final int PADDING_FROM_ARGUMENT = -2;
```

Comment:
Accidentally deleted a comment from Benoit here saying it's using in `Time.httpdate` and that there should be a spec for it.